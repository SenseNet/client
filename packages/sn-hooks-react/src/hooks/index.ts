export * from './use-authentication'
export * from './use-download'
export * from './use-injector'
export * from './use-logger'
export * from './use-repository-events'
export * from './use-repository'
export * from './use-session'
export * from './use-version-info'
export * from './use-wopi'
