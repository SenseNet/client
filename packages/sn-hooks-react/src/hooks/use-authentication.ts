import { authenticationResultStatus, AuthFuncResult, OIDCAuthenticationService } from '@sensenet/client-core'
import { useState } from 'react'
import { useLogger } from '.'

export interface AuthActionParams {
  returnUrl: string
  authService: OIDCAuthenticationService
  successCallback?: (url: string) => void
}

/**
 * useAuthentication hook helps provide helper methods for login and logout with an OIDC authentication service.
 * It **doesn't** work with forms authentication.
 */
export function useAuthentication() {
  const [isLoading, setIsLoading] = useState(false)
  const [error, setError] = useState<string>()
  const logger = useLogger('useAuthentication')

  const doAuthenticationAction = async (action: () => AuthFuncResult, successCallback?: (url: string) => void) => {
    setIsLoading(true)
    const result = await action()
    setIsLoading(false)
    switch (result.status) {
      case authenticationResultStatus.redirect:
        break
      case authenticationResultStatus.success:
        successCallback?.(result.state.returnUrl)
        break
      case authenticationResultStatus.fail:
        logger.warning({ message: result.error.message })
        setError(result.error.message)
        break
      default:
        logger.debug({ message: `Invalid status result ${result}.` })
    }
  }

  const login = async ({ authService, returnUrl, successCallback }: AuthActionParams) => {
    await doAuthenticationAction(() => authService.signIn({ returnUrl }), successCallback)
  }

  const logout = async ({ authService, returnUrl, successCallback }: AuthActionParams) => {
    await doAuthenticationAction(() => authService.signOut({ returnUrl }), successCallback)
  }

  return { login, isLoading, error, logout }
}
