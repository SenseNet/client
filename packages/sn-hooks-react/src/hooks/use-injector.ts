import { useContext } from 'react'
import { InjectorContext } from '../context'

/**
 * Custom hook that returns an Injecor instance from the InjectorContext.
 */
export const useInjector = () => useContext(InjectorContext)
