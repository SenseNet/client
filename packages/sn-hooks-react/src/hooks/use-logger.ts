import { useContext, useState } from 'react'
import { LoggerContext } from '../context/LoggerContext'

/**
 * Returns a scoped Logger instance from the LoggerContext.
 * @param scope The logger's Scope name
 */
export const useLogger = (scope: string) => {
  const [logger] = useState(useContext(LoggerContext).withScope(scope))
  return logger
}
