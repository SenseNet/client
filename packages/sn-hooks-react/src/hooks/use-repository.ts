import { useContext } from 'react'
import { RepositoryContext } from '../context'

/**
 * Returns a Repository instance from the RepositoryContext
 */
export const useRepository = () => useContext(RepositoryContext)
