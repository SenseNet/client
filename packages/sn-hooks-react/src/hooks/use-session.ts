import { useContext } from 'react'
import { SessionContext } from '../context'

export const useSession = () => useContext(SessionContext)
