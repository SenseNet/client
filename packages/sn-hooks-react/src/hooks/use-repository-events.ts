import { useEffect, useState } from 'react'
import { EventHub } from '@sensenet/repository-events'
import { useRepository } from './use-repository'

export const useRepositoryEvents = () => {
  const repo = useRepository()
  const [eventHub, setEventHub] = useState(new EventHub(repo))

  useEffect(() => {
    const eh = new EventHub(repo)
    setEventHub(eh)
    return () => eh.dispose()
  }, [repo])

  return eventHub
}
