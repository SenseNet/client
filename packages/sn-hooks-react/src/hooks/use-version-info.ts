import { ConstantContent } from '@sensenet/client-core'
import { useEffect, useState } from 'react'
import { useRepository } from './use-repository'

/*
 * The type of a specific package
 */
export enum PackageType {
  /**
   * Can contain small repeatable activities that do not perform significant changes but can be important because of business or technical reasons. A good example is performing an undo checkout on multiple content. Executing a package of this level does not change the application's or the product's version number but the execution is logged and registered
   */
  Tool,

  /**
   * Contains small modifications (e.g. a couple of new content to import or a bugfix in a dll). Usually patches form a chain where every package assumes the existence of all the previous ones but it is not mandatory. It is possible to control this behavior, see version control below.
   */
  Patch,

  /**
   *  An application's first package must be an 'install' package. This is the package that injects a new application into the system. Only Application packages can be set on this level. An install level package must contain a new application identifier that is unknown to the system. Packages on this level can be executed only once.
   */
  Install,
}

/**
 * Represents a .NET Assembly in the Version Info
 */
export interface Assembly {
  CodeBase: string
  IsDynamic: boolean
  Name: string
  Version: string
}

/**
 * Represents a sensenet Component in the Version Info
 */
export interface Component {
  ComponentId: string
  Version: string
  AcceptableVersion: string
  Description: string
  IsUpdateAvailable?: boolean
  NugetManifest?: any
}

/**
 * Represents a sensenet Package in the Version Info
 */
export interface Package {
  ComponentId: string
  ComponentVersion: string
  Description: string
  ExecutionDate: Date
  ExecutionError: any
  Id: number
  Manifest: any
  PackageType: PackageType
  ReleaseDate: Date
}

/**
 * Represents a model for the sensenet's GetVersionInfo custom action's response
 */
export interface VersionInfo {
  Assemblies: {
    Dynamic: Assembly[]
    GAC: Assembly[]
    Other: Assembly[]
    Plugins: Assembly[]
    SenseNet: Assembly[]
  }
  Components: Component[]
  DatabaseAvailable: boolean
  InstalledPackages: Package[]
}

/**
 * Returns the current repository's Version Info values and also checks if the installed components has an updated version on https://nuget.org
 */
export const useVersionInfo = () => {
  const [versionInfo, setVersionInfo] = useState<VersionInfo | undefined>()
  const [nugetManifests, setNugetManifests] = useState<any[]>([])
  const [hasUpdates, setHasUpdates] = useState(false)
  const repo = useRepository()

  useEffect(() => {
    ;(async () => {
      const result = await repo.executeAction<undefined, VersionInfo>({
        idOrPath: ConstantContent.PORTAL_ROOT.Path,
        body: undefined,
        method: 'GET',
        name: 'GetVersionInfo',
      })

      const nugetPromises = result.Components.map(async (component) => {
        try {
          const response = await fetch(
            `https://api.nuget.org/v3/registration3-gz-semver2/${component.ComponentId.toLowerCase()}/index.json`,
          )
          if (response.ok) {
            const nugetManifest = await response.json()
            return nugetManifest
          }
        } catch (error) {
          return {}
        }
      })
      const loadedManifests = (await Promise.all(nugetPromises)).filter((m) => m)
      setNugetManifests(loadedManifests)

      let hasOneUpdate = false

      result.Components = result.Components.map((component) => {
        const nugetManifest = loadedManifests.find(
          (m) =>
            m['@id'] ===
            `https://api.nuget.org/v3/registration3-gz-semver2/${component.ComponentId.toLocaleLowerCase()}/index.json`,
        )
        const updateAvailable = nugetManifest ? nugetManifest.items[0].upper > component.Version : false
        if (updateAvailable) {
          hasOneUpdate = true
        }
        return {
          ...component,
          NugetManifest: nugetManifest,
          IsUpdateAvailable: updateAvailable,
        }
      })
      setHasUpdates(hasOneUpdate)
      setVersionInfo(result)
    })()
  }, [repo])

  return { versionInfo, nugetManifests, hasUpdates }
}
