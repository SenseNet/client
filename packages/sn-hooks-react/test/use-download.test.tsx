import React from 'react'
import { GenericContent } from '@sensenet/default-content-types'
import { mount } from 'enzyme'
import * as downloadHook from '../src/hooks/use-download'

const DownloadDump: React.FC<{ content: GenericContent }> = ({ content }) => {
  const d = downloadHook.useDownload(content)
  return <div onClick={d.download}>IsFile: {d.isFile}</div>
}

describe('useDownload()', () => {
  it('for non-file contents', () => {
    const t = mount(
      <DownloadDump
        content={{
          Id: 1,
          Type: 'User',
          Name: '',
          Path: '',
        }}
      />,
    )
    expect(t).toMatchSnapshot()
  })

  it('for file contents', () => {
    const t = mount(
      <DownloadDump
        content={{
          Id: 1,
          Type: 'File',
          Name: '',
          Path: '',
        }}
      />,
    )
    expect(t).toMatchSnapshot()
  })

  it('should trigger download on click', () => {
    jest.spyOn(downloadHook, 'downloadFile')
    jest.spyOn(downloadHook, 'fakeClick')
    const t = mount(
      <DownloadDump
        content={{
          Id: 1,
          Type: 'File',
          Name: '',
          Path: '',
        }}
      />,
    )
    ;(t.find('div').prop('onClick') as any)()
    expect(downloadHook.downloadFile).toBeCalled()
    expect(downloadHook.fakeClick).toBeCalled()
  })
})
