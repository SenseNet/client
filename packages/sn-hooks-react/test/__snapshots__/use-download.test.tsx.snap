// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDownload() for file contents 1`] = `
<DownloadDump
  content={
    Object {
      "Id": 1,
      "Name": "",
      "Path": "",
      "Type": "File",
    }
  }
>
  <div
    onClick={[Function]}
  >
    IsFile: 
  </div>
</DownloadDump>
`;

exports[`useDownload() for non-file contents 1`] = `
<DownloadDump
  content={
    Object {
      "Id": 1,
      "Name": "",
      "Path": "",
      "Type": "User",
    }
  }
>
  <div
    onClick={[Function]}
  >
    IsFile: 
  </div>
</DownloadDump>
`;
